stock_request/models/stock_request_order.py

```diff
@@ -0,0 +1,359 @@
+
+    @api.onchange("procurement_group_id")
+    def onchange_procurement_group_id(self):
+        self.change_childs()
+
+    @api.onchange("company_id")
+    def onchange_company_id(self):
+        if self.company_id and (
+            not self.warehouse_id or self.warehouse_id.company_id != self.company_id
+        ):
+            self.warehouse_id = self.env["stock.warehouse"].search(
+                [("company_id", "=", self.company_id.id)], limit=1
+            )
+            self.with_context(no_change_childs=True).onchange_warehouse_id()
+        self.change_childs()
+
+    def change_childs(self):
+        if not self._context.get("no_change_childs", False):
+            for line in self.stock_request_ids:
+                line.warehouse_id = self.warehouse_id
+                line.location_id = self.location_id
+                line.company_id = self.company_id
+                line.picking_policy = self.picking_policy
+                line.expected_date = self.expected_date
+                line.requested_by = self.requested_by
+                line.procurement_group_id = self.procurement_group_id
+
+    def action_confirm(self):
+        if not self.stock_request_ids:
+            raise UserError(
+                _("There should be at least one request item for confirming the order.")
+            )
+        self.mapped("stock_request_ids").action_confirm()
+        return True
+
+    def action_draft(self):
+        self.mapped("stock_request_ids").action_draft()
+        return True
+
+    def action_cancel(self):
+        self.mapped("stock_request_ids").action_cancel()
+        return True
+
+    def action_done(self):
+        return True
+
+    def action_view_transfer(self):
+        action = self.env["ir.actions.act_window"]._for_xml_id(
+            "stock.action_picking_tree_all"
+        )
+
+        pickings = self.mapped("picking_ids")
+        if len(pickings) > 1:
+            action["domain"] = [("id", "in", pickings.ids)]
+        elif pickings:
+            action["views"] = [(self.env.ref("stock.view_picking_form").id, "form")]
+            action["res_id"] = pickings.id
+        return action
+
+    def action_view_stock_requests(self):
+        action = self.env["ir.actions.act_window"]._for_xml_id(
+            "stock_request.action_stock_request_form"
+        )
+        if len(self.stock_request_ids) > 1:
+            action["domain"] = [("order_id", "in", self.ids)]
+        elif self.stock_request_ids:
+            action["views"] = [
+                (self.env.ref("stock_request.view_stock_request_form").id, "form")
+            ]
+            action["res_id"] = self.stock_request_ids.id
+        return action
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get("name", "/") == "/":
+                vals["name"] = self.env["ir.sequence"].next_by_code(
+                    "stock.request.order"
+                )
```

Comment:
Say for instance that the sequence is configured as no gap. It will be rolled back on retry and the retry will then use names that are not consumed in the sequence, leading to duplicate names. 

So in any case, modifying vals in create and write is very fragile and this should be strongly discouraged.

Reply:
> Too many if now for triggering the problem. 

Sure but bugs always happens when the code is used in initially unforeseen ways and that's why one must program defensively without assumptions on how a function will be used.

> I for example discourage to use no gap sequences.

Me too, but they exist.

> This should be solved at ORM level, not at programmer level. 

I agree but until then I'll raise these every time I see them in code review.

> I repeat the question: have you talked with Odoo about this? Are they reluctant to change?

I think so yes, I don't remember exactly. I've never created an issue though. But if *you* raise the performance argument, you can be sure that Odoo will raise it as a reason not to do the change :grin: The only way to do it while not impacting performance is to deserialize as immutable data structures, but that's going to be considered a breaking change.

Reply:
Yes, there are less chances to get this fixed taking the path of frozendicts, and this also requires a copy to that structure, so I think the solution of copying arguments is reasonable, as the copy is only done on the layer of receiving the RPC calls. I will raise the issue in Odoo pinging you.